### PR TITLE
Update arrow version to fix plasma bugs

### PR DESCRIPTION
**Pull request by guoyuhong**
## What do these changes do?
This PR updates arrow version to b22848952f09d6f9487feaff80ee358ca41b1562 which contains https://github.com/apache/arrow/pull/3656 and https://github.com/apache/arrow/pull/3682 and this PR will fix 2 TODOs from Plasma Java Client.

## Related issue number
N/A

### CODE REVIEW

**Review by raulchen — APPROVED** [review]
LGTM

**Comment by robertnishihara** [comment]
@guoyuhong Did this work for you? @pcmoritz is also updating the Arrow version in https://github.com/ray-project/ray/pull/3898 and it's gotten a lot more complex because the Arrow build system has changed.

**Comment by guoyuhong** [comment]
@pcmoritz I rebased the original ref branch [arrow-one-mmap-file](https://github.com/ray-project/arrow/commit/007e1ca289e979bac80231fa9ee7510be744b60b) to the new base of b22848952f09d6f9487feaff80ee358ca41b1562 and resolve the conflicts and push it to [arrow-one-mmap-file-on-b228489](https://github.com/ray-project/arrow/commit/54618466ca04f4427dd8cfc3ce31b14ec54e8999). Is the processes right?

**Comment by guoyuhong** [comment]
@robertnishihara His PR seems not update the Arrow version. Do you mean that we hold the PR until that PR gets merged?

**Comment by guoyuhong** [comment]
@robertnishihara That PR seems to use Arrow directly from OSS...

**Comment by robertnishihara** [comment]
@guoyuhong I believe it uses Arrow from an S3 bucket that we host. We have to separately build Arrow using https://github.com/ray-project/arrow-build I think, which uploads to the S3 bucket and then we can get it from the bucket. This is done to address issues @pcmoritz ran into when trying to get the Bazel build to work with Arrow.

**Comment by guoyuhong** [comment]
@robertnishihara When will @pcmoritz 's PR get merged? This PR contains 2 fixes along with the Plasma Java Client, so this PR cannot be closed.

**Comment by robertnishihara** [comment]
@guoyuhong I'm hoping to merge @pcmoritz's PR in the morning (in about 8 hours). It's basically ready, but there were a couple small changes that needed to be made.

**Comment by guoyuhong** [comment]
@robertnishihara Thanks for the information. Let's hold this PR until then.

**Comment by guoyuhong** [comment]
@robertnishihara Considering that the Bazel PR is not merged, I think I can send a PR to update https://github.com/ray-project/arrow-build 's Arrow version to 54618466ca04f4427dd8cfc3ce31b14ec54e8999 which is the same as this PR and generate the new wheel files. Then, @pcmoritz 's PR will compatible with this PR using the new wheel files.

**Comment by pcmoritz** [comment]
@guoyuhong Thanks for looking into this, see my comment in https://github.com/ray-project/arrow-build/pull/1

I have now a plan how we can go forward on https://github.com/ray-project/ray/issues/4129 and make sure the latest arrow features are available and the clashes with tensorflow are removed. It involves building an arrow wheel that uses ubuntu 14.04 as a base (the same as tensorflow) and I'll prioritize this first (it will allow us to remove CMake) and can help you with this PR after.

**Comment by guoyuhong** [comment]
@pcmoritz https://github.com/ray-project/ray/pull/4182 had been merged. How about this PR? Since Java building still goes the cmake way,  we need to update is correspondingly. It is better to keep the version of cmake and bazel version the same.

**Comment by robertnishihara** [comment]
@guoyuhong, this PR updates the version of Arrow used by cmake to the commit b22848952f09d6f9487feaff80ee358ca41b1562 which is identical to a commit in the Arrow repository. That means that it does not have the patch https://github.com/ray-project/arrow/pull/3, which was needed to make Ray and TensorFlow compatible.

The tests are passing in this PR probably because when they ran, we were using `conda install tensorflow` in the Travis tests. However, the current master uses `pip install tensorflow` in the Travis tests and so I think it may fail (I rebased the PR to check).

If we merge this, we need to use a commit that includes the patch https://github.com/ray-project/arrow/pull/3.

**Comment by guoyuhong** [comment]
@robertnishihara Sorry for the expression. What I mean is that this PR will update to the version that  https://github.com/ray-project/arrow/pull/3 rebased on 
b22848952f09d6f9487feaff80ee358ca41b1562. And the commit is [54618466ca04f4427dd8cfc3ce31b14ec54e8999](https://github.com/ray-project/arrow/commit/54618466ca04f4427dd8cfc3ce31b14ec54e8999).

**Comment by guoyuhong** [comment]
I will do a rebase for this PR to current master.

**Comment by robertnishihara** [comment]
Oh, I apologize, I meant to check the commit from the PR but I guess I checked the one from the commit message. My mistake!

**Comment by pcmoritz** [comment]
Great @guoyuhong thanks for the PR! If it passes on the rebase, let's include it in the point release we are about to finish! Let's however hold off merging big PRs that touch the backend before the release candidate is drafted.

**Comment by guoyuhong** [comment]
There is build error in Arrow CMake and it is fixed by https://github.com/apache/arrow/pull/3737 . I need to rebase the commit to this commit.

**Comment by guoyuhong** [comment]
I changed the commit to https://github.com/ray-project/arrow/commit/2c511979b13b230e73a179dab1d55b03cd81ec02 of branch [arrow-one-mmap-file-on-46f75d7](https://github.com/ray-project/arrow/commits/arrow-one-mmap-file-on-46f75d7) which is based on the cmake fix of https://github.com/apache/arrow/pull/3737 . Let's wait for the CI results.

**Comment by robertnishihara** [comment]
Jenkins, retest this please.

**Comment by robertnishihara** [comment]
In

`++ docker run --rm --shm-size=20G --memory=20G 9d42513dc5bb6f3116c18c7e4312abc0219e6c793053b4db4b7d56b451b04346 /ray/python/ray/rllib/test/run_silent.sh train.py --env CartPole-v0 --run MARWIL --stop '{"training_iteration": 2}' --config '{"input": "/ray/python/ray/rllib/test/data/cartpole_small", "learning_starts": 0, "input_evaluation": ["wis", "is"], "shuffle_buffer_size": 10}'` 

I see

```
(pid=68) 2019-03-01 11:03:10,496	INFO policy_evaluator.py:275 -- Creating policy evaluation worker 0 on CPU (please ignore any CUDA init errors)
(pid=68) 2019-03-01 11:03:10.507560: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2593605000 Hz
(pid=68) 2019-03-01 11:03:10.510056: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x558942f5d300 executing computations on platform Host. Devices:
(pid=68) 2019-03-01 11:03:10.510102: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
(pid=68) WARNING:tensorflow:From /opt/conda/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
(pid=68) Instructions for updating:
(pid=68) Colocations handled automatically by placer.
(pid=68) WARNING:tensorflow:From /opt/conda/lib/python2.7/site-packages/tensorflow/python/util/decorator_utils.py:145: VARIABLES (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
(pid=68) Instructions for updating:
(pid=68) Use `tf.GraphKeys.GLOBAL_VARIABLES` instead.
(pid=68) WARNING:tensorflow:From /ray/python/ray/rllib/models/action_dist.py:114: multinomial (from tensorflow.python.ops.random_ops) is deprecated and will be removed in a future version.
(pid=68) Instructions for updating:
(pid=68) Use tf.random.categorical instead.
(pid=68) 2019-03-01 11:03:10,894	WARNING json_reader.py:52 -- Treating input directory as glob pattern: /ray/python/ray/rllib/test/data/cartpole_small/*.json
(pid=68) 2019-03-01 11:03:10,895	INFO json_reader.py:65 -- Found 1 input files.
(pid=68) 2019-03-01 11:03:10,901	INFO shuffled_input.py:39 -- Filling shuffle buffer to 10 batches
(pid=68) terminate called after throwing an instance of 'std::invalid_argument'
(pid=68)   what():  stoi
(pid=68) Fatal Python error: Aborted
(pid=68) 
(pid=68) Stack (most recent call first):
(pid=68)   File "/ray/python/ray/rllib/utils/compression.py", line 49 in unpack
(pid=68)   File "/ray/python/ray/rllib/utils/compression.py", line 56 in unpack_if_needed
(pid=68)   File "/ray/python/ray/rllib/offline/json_reader.py", line 153 in _from_json
(pid=68)   File "/ray/python/ray/rllib/offline/json_reader.py", line 105 in _try_parse
(pid=68)   File "/ray/python/ray/rllib/offline/json_reader.py", line 72 in next
(pid=68)   File "/ray/python/ray/rllib/offline/shuffled_input.py", line 41 in next
(pid=68)   File "/ray/python/ray/rllib/evaluation/policy_evaluator.py", line 389 in sample
(pid=68)   File "/ray/python/ray/rllib/optimizers/sync_batch_replay_optimizer.py", line 53 in step
(pid=68)   File "/ray/python/ray/rllib/agents/marwil/marwil.py", line 65 in _train
(pid=68)   File "/ray/python/ray/tune/trainable.py", line 150 in train
(pid=68)   File "/ray/python/ray/rllib/agents/agent.py", line 293 in train
(pid=68)   File "/ray/python/ray/function_manager.py", line 783 in actor_method_executor
(pid=68)   File "/ray/python/ray/worker.py", line 857 in _process_task
(pid=68)   File "/ray/python/ray/worker.py", line 958 in _wait_for_and_process_task
(pid=68)   File "/ray/python/ray/worker.py", line 1007 in main_loo2019-03-01 11:03:11,052	ERROR trial_runner.py:426 -- Error processing event.
Traceback (most recent call last):
  File "/ray/python/ray/tune/trial_runner.py", line 389, in _process_events
    result = self.trial_executor.fetch_result(trial)
  File "/ray/python/ray/tune/ray_trial_executor.py", line 231, in fetch_result
    result = ray.get(trial_future[0])
  File "/ray/python/ray/worker.py", line 2275, in get
    raise value
RayActorError: The actor died unexpectedly before finishing this task.
2019-03-01 11:03:11,055	ERROR worker.py:1737 -- A worker died or was killed while executing task 00000000fc55b69c2f8dc56cc5a863944b94a8d7.
```

Does anyone know if this error is new or not?

**Comment by robertnishihara** [comment]
Jenkins, retest this please.

**Comment by robertnishihara** [comment]
The `std::invalid_argument` might be a new error.

**Comment by guoyuhong** [comment]
This PR may help to resolve this problem. https://github.com/apache/arrow/pull/3758

**Comment by guoyuhong** [comment]
I changed the commit to https://github.com/ray-project/arrow/commit/68299c5f48289c4f39a948cbd0426b9199a9df1e of branch [arrow-one-mmap-file-on-c0a2e73](https://github.com/ray-project/arrow/commits/arrow-one-mmap-file-c0a2e73) which is based on the cmake fix of https://github.com/apache/arrow/pull/3758 .

**Comment by guoyuhong** [comment]
@ericl Since https://github.com/apache/arrow/pull/3423 and the fix https://github.com/apache/arrow/pull/3758 have been merged, the arrow serialization is different. There is a checked-in file https://github.com/ray-project/ray/blob/master/python/ray/rllib/test/data/cartpole_small/output-2019-02-03_20-27-20_worker-0_0.json which is used in Jenkins test but throws exception during deserialization when I try to update arrow version. All the Travis tests passed. Could you help to update that? CC: @pcmoritz

**Comment by guoyuhong** [comment]
@ericl Ping... Could you help to take a look?

**Comment by ericl** [comment]
I see, feel free to disable the test. I can fix it in a follow-up PR.

File: ci/jenkins_tests/run_rllib_tests.sh
Comment (guoyuhong) [inline ci/jenkins_tests/run_rllib_tests.sh]: @ericl I disabled 3 cases in this script and left a TODO comment for you. Now the Jenkins test has no such failures.

**Comment by guoyuhong** [comment]
Hi all, I have rebased the PR and disable the Jenkins cases that caused by the Arrow serialization change which @ericl promised to fix later. Please help to review.

**Review by raulchen — APPROVED** [review]
LGTM

**Comment by robertnishihara** [comment]
Looks fine to me assuming tests pass. There is some merge conflict.

**Pull request merged**
